src/soter/openssl/soter_sign_ecdsa.c

```diff
@@ -31,74 +31,61 @@ soter_status_t soter_sign_init_ecdsa_none_pkcs8(soter_sign_ctx_t* ctx,
                                                 const size_t public_key_length)
 {
     soter_status_t err = SOTER_FAIL;
-    EVP_PKEY* pkey = NULL;
 
-    pkey = EVP_PKEY_new();
-    if (!pkey) {
-        return SOTER_NO_MEMORY;
-    }
-
-    if (!EVP_PKEY_set_type(pkey, EVP_PKEY_EC)) {
-        goto free_pkey;
-    }
-
-    ctx->pkey_ctx = EVP_PKEY_CTX_new(pkey, NULL);
-    if (!(ctx->pkey_ctx)) {
-        err = SOTER_NO_MEMORY;
-        goto free_pkey;
-    }
-
-    if (!EVP_PKEY_paramgen_init(ctx->pkey_ctx)) {
-        goto free_pkey_ctx;
-    }
-    if (!EVP_PKEY_CTX_set_ec_paramgen_curve_nid(ctx->pkey_ctx, NID_X9_62_prime256v1)) {
-        goto free_pkey_ctx;
-    }
-    if (!EVP_PKEY_paramgen(ctx->pkey_ctx, &pkey)) {
-        goto free_pkey_ctx;
+    /* soter_sign_ctx_t should be initialized only once */
+    if (!ctx || ctx->pkey || ctx->md_ctx) {
+        return SOTER_INVALID_PARAMETER;
     }
 
     if ((!private_key) && (!public_key)) {
-        err = soter_ec_gen_key(ctx->pkey_ctx);
+        err = soter_ec_gen_key(&ctx->pkey);
         if (err != SOTER_SUCCESS) {
-            goto free_pkey_ctx;
+            goto free_pkey;
         }
     } else {
+        ctx->pkey = EVP_PKEY_new();
+        if (!ctx->pkey) {
+            err = SOTER_NO_MEMORY;
+            goto free_pkey;
+        }
+
+        if (!EVP_PKEY_set_type(ctx->pkey, EVP_PKEY_EC)) {
+            err = SOTER_FAIL;
+            goto free_pkey;
+        }
+
         if (private_key != NULL) {
-            err = soter_ec_import_key(pkey, private_key, private_key_length);
+            err = soter_ec_import_key(ctx->pkey, private_key, private_key_length);
             if (err != SOTER_SUCCESS) {
-                goto free_pkey_ctx;
+                goto free_pkey;
             }
         }
         if (public_key != NULL) {
-            err = soter_ec_import_key(pkey, public_key, public_key_length);
+            err = soter_ec_import_key(ctx->pkey, public_key, public_key_length);
             if (err != SOTER_SUCCESS) {
-                goto free_pkey_ctx;
+                goto free_pkey;
             }
         }
     }
 
     ctx->md_ctx = EVP_MD_CTX_create();
     if (!(ctx->md_ctx)) {
         err = SOTER_NO_MEMORY;
-        goto free_pkey_ctx;
+        goto free_pkey;
     }
 
-    if (EVP_DigestSignInit(ctx->md_ctx, NULL, EVP_sha256(), NULL, pkey) != 1) {
+    if (EVP_DigestSignInit(ctx->md_ctx, NULL, EVP_sha256(), NULL, ctx->pkey) != 1) {
         goto free_md_ctx;
     }
 
-    EVP_PKEY_free(pkey);
     return SOTER_SUCCESS;
```

Comment:
shouldn't we free anything temp before returning success?

Reply:
> shouldn't we free anything temp before returning success?

It's not a temporary though, this EVP_PKEY is owned by `ctx`.